library/n2_protocol.hpp

```diff
@@ -0,0 +1,123 @@
+#ifndef ELLIPTICS_N2_PROTOCOL_HPP
+#define ELLIPTICS_N2_PROTOCOL_HPP
+
+#include <optional>
+
+#include <elliptics/packet.h>
+#include <elliptics/utils.hpp>
+
+namespace ell = ioremap::elliptics;
+
+namespace ioremap { namespace elliptics { namespace n2 {
+
+using cmd_native = dnet_cmd;
+
+struct data_in_file {
+	int fd = -1;
+	off_t local_offset = 0;
+	size_t fsize = 0;
+	int on_exit = 0;
+};
+
+struct data_place {
+	enum place_type {
+		IN_MEMORY,
+		IN_FILE,
+	};
+
+	data_in_file in_file;
+	ell::data_pointer in_memory;
+
+	place_type where() const;
+
+	static data_place from_file(const data_in_file &in_file);
+	static data_place from_memory(const ell::data_pointer &in_memory);
+};
+
+class isender;
```

Comment:
There is no implementation for `isender`. Should it be?

Reply:
I expect that @artsel implement it. But I ask him, maybe he doesn't want to use this pattern.